clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/deploy/util/id/DcosSpinnakerAppId.groovy

```diff
@@ -0,0 +1,144 @@
+        marathonPath.first().get()
+    }
+
+    /**
+     * @return The canonical DC/OS "region" (a.k.a the full group path in which the marathon application lives)
+     *         including backslashes. This is returned as a relative path, meaning no preceeding backslash. Will never
+     *         be null.
+     *         <p/>
+     *         Deemed unsafe because various Spinnaker components have trouble with a region with backslashes.
+     *         <p/>
+     *         Ex: {@code foo/bar}
+     * @see #getSafeRegion()
+     */
+    public String getRegion() {
+        marathonPath.tail().parent().relative()
+    }
+
+    /**
+     * @return The canonical DC/OS "region" (a.k.a the full group path in which the marathon application lives)
+     *         including backslashes. This is returned as a relative path, meaning no preceeding backslash. Will never
+     *         be null.
+     *         <p/>
+     *         Deemed unsafe because various Spinnaker components have trouble with a region with backslashes.
+     *         <p/>
+     *         Ex: {@code foo/bar}
+     * @see #getSafeRegion()
+     */
+    public String getUnsafeRegion() {
+        region
+    }
+
+    /**
+     * @return The "safe" DC/OS region (a.k.a the group in which the marathon application lives). This is returned as a
+     *         relative path, meaning no preceeding underscore. Will never be null.
+     *         <p/>
+     *         Deemed safe because backslashes are replaced with underscores.
+     *         <p/>
+     *         Ex: {@code acct_foo_bar}
+     * @see #getSafeRegion()
+     */
+    public String getSafeRegion() {
+        region.replaceAll(MarathonPathId.PART_SEPARATOR, SAFE_REGION_SEPARATOR)
+    }
+
+    public Names getServerGroupName() {
+        Names.parseName(marathonPath.last().get())
+    }
+
+    public MarathonPathId toMarathonPathId() {
+        marathonPath
+    }
+
+    @Override
+    public String toString() {
+        marathonPath.toString()
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true
+        if (o == null || getClass() != o.getClass())
+            return false
+        def dcosPathId = (DcosSpinnakerAppId) o
+        return dcosPathId.toString() == toString()
+    }
+
+    @Override
+    public int hashCode() {
+        return toString().hashCode()
+    }
+
+    private static void validateServerGroupName(final String serverGroupName) {
+        def service = Names.parseName(serverGroupName)
+
+        checkArgument(!nullToEmpty(service.app).trim().empty, "The server group app should not be null, empty, or blank.")
+        checkArgument(service.sequence >= 0, "The server group sequence should not be negative or null.")
+    }
+
+    public static Optional<DcosSpinnakerAppId> from(String marathonAppId) {
+        try {
+            Optional.of(new DcosSpinnakerAppId(marathonAppId))
+        } catch (IllegalArgumentException e) {
```

Comment:
It doesn't look like invalid ids would get created (https://github.com/cerner/clouddriver/pull/10/files#diff-589ac0bedd6bcd7cb554c06ef4e59e27L54). There was a stipulation in the doc that you should ensure the id is valid first (otherwise it would throw an exception, not create the invalid id)

However, I think encapsulating that in the static factory method is good (with the Optional return), though I think only the "parse" factory methods (i.e. the ones that take in a marathonAppId) should have it. Doesn't seem necessary for the ones that take in discreet parts, since all the calling sites should already have valid values for all of those.